Move Object variables to Object Editor Dialog

Midhil457 opened the pull request:
Moved Object variables to Object Editor Dialog (#3244)

- [x] Added object variables to object editor
- [x] Update `Edit object variable` action to direct to the new tab
- [x] (Tried) Remove unnessesary code
- [x] Fix VariableList component not taking up all the area

I need help with the last one, the component seems to be moved to the side. I have spent a few days trying to fix it, with no good result (Most likely due to my lack of experience :sweat_smile: ) , it would be great if someone could help find and fix the problem.

EDIT: Fixed it, see next comment
Everything else seems to work perfectly!

![gdghalf](https://user-images.githubusercontent.com/73597906/142652138-c0c206a2-5480-4370-a9a2-ca455a26877e.png)

:)

Midhil457 commented:
I was able to fix the issue by using a Column and adding `expand` and `noMargin` :partying_face:
![fixvar](https://user-images.githubusercontent.com/73597906/142716726-e35d067e-1f6f-4afe-8162-f536fdd38096.png)

4ian commented on newIDE/app/src/ObjectEditor/ObjectEditorDialog.js:
```diff
@@ -188,6 +196,36 @@ const InnerDialog = (props: InnerDialogProps) => {
+        onSizeUpdated={
+          forceUpdate /*Force update to ensure dialog is properly positioned*/
+        }
+        onComputeAllVariableNames={() =>
+          EventsRootVariablesFinder.findAllObjectVariables(
+            props.project.getCurrentPlatform(),
+            props.project,
+            props.layout,
+            props.object
+        )}
```
props.layout will never exist/be set, so this will never work properly. I think we can keep this call to findAllObjectVariables in SceneEditor/index.js and pass it as a prop:

```suggestion
        onComputeAllVariableNames={props.onComputeAllVariableNames}
```

and in SceneEditor/index.js:

```
<ObjectEditorDialog
  ...
  onComputeAllVariableNames={() => EventsRootVariablesFinder.findAllObjectVariables(...)}
/>
```

This way we avoid polluting ObjectEditorDialog by passing it the layout - which would be only useful for this and so would be meaning passing "too much information" to the component while it should not have access to the layout.

4ian commented on newIDE/app/src/ObjectEditor/ObjectEditorDialog.js:
```diff
@@ -23,10 +23,13 @@ import HotReloadPreviewButton, {
 type Props = {|
   open: boolean,
   object: ?gdObject,
+  variables: ?gdObject.variables,
```
This error is also reported by Flow, the type checker, that you can find in the CI runs: https://gdevelop.semaphoreci.com/jobs/c3aa34e5-238e-4c65-b021-87a1ed661b32

4ian commented on newIDE/app/src/ObjectsList/index.js:
```diff
@@ -10,7 +10,6 @@ import SortableVirtualizedItemList from '../UI/SortableVirtualizedItemList';
 import Background from '../UI/Background';
 import SearchBar from '../UI/SearchBar';
 import NewObjectDialog from '../AssetStore/NewObjectDialog';
```
Changes in the file seem all good 👍

4ian commented on newIDE/app/src/VariablesList/VariablesEditorDialog.js:
```diff
@@ -63,14 +62,6 @@ const VariablesEditorDialog = ({
       cannotBeDismissed={true}
       onRequestClose={onCancelChanges}
       secondaryActions={[
```
You've removed a feature here, this means we can't open the object variables anymore when we edit the variables of an instance.
Add it back and find a way instead to open the object editor with the variable tab opened :) 

This means you'll have to see where this is used, search for `onEditObjectVariables` in the codebase.
You'll find it in SceneEditor/index.js (ignore the things related to `UseSceneEditorCommands`, it's a totally different component), here precisely:

https://github.com/4ian/GDevelop/blob/2517b4740111a3dc790f27f6c7b4d671b26a0f1d/newIDE/app/src/SceneEditor/index.js#L1315-L1331

If you read this, you see this opens a variable editor for the object. You need to:
1) Replace this by a call to `this.editObject(object, 'variables')` - simple! :) 
2) A bit harder: now you must clean all usage of `this.state.variablesEditedObject` and remove `<VariablesEditorDialog` which is related to `this.state.variablesEditedObject`. Why? Because this dialog is now useless, as we moved the variables edition inside the object editor :)

4ian reviewed (commented):
That's a good start! Thanks for taking the time to do this 👍

See my comments, there are a few points notably one to:
- Put back the button to open object variables in the instance variables (careful when doing this, you must call this explicitly in your PR description because otherwise we risk losing features without noticing)
- Adapt the SceneEditor/index.js which was also containing an object variable editor, that we can remove.

Thanks! :)

Midhil457 commented:
Thank you for the review!
I have made the changes according to the review :)

One change I want to make but need the green flag is on replacing the empty message with the `EmptyPlaceHolder` (If that's the name of it) to match the effects and behavior tabs.

Should the `+ Add` button be moved to the bottom? To match the other tab's style. It wouldn't adversly affect the VariableDialog much. I don't exactly know how to do it though (I did try it a few times) :sweat_smile: So I would appreciate some help

4ian commented:
> One change I want to make but need the green flag is on replacing the empty message with the EmptyPlaceHolder (If that's the name of it) to match the effects and behavior tabs.

I think that would be a good idea! 
Feel free to go ahead and do this. Make sure to also test the result in the case of the scene or global variable editor :) 

> Should the + Add button be moved to the bottom? To match the other tab's style. It wouldn't adversly affect the VariableDialog much. I don't exactly know how to do it though (I did try it a few times) 😅 So I would appreciate some help

We can go back to this later, for now it's still ok this way! :)

4ian commented on newIDE/app/src/SceneEditor/index.js:
```diff
@@ -1194,6 +1194,26 @@ export default class SceneEditor extends React.Component<Props, State> {
             resourceExternalEditors={resourceExternalEditors}
             onChooseResource={onChooseResource}
+            onComputeAllVariableNames={() => {
+              const variablesEditedInstance = this.state
+                .variablesEditedInstance;
+              if (!variablesEditedInstance) {
+                return [];
+              }
+              const variablesEditedObject = getObjectByName(
+                project,
+                layout,
+                variablesEditedInstance.getObjectName()
+              );
+              return variablesEditedObject
+                ? EventsRootVariablesFinder.findAllObjectVariables(
+                    project.getCurrentPlatform(),
+                    project,
+                    layout,
+                    variablesEditedObject
+                  )
+                : [];
```
This seems wrong, you check the object of the edited instance, but we're in the object editor, so we're editing an object.

instead, just pass `this.state.editedObjectWithContext.object` to the last argument of findAllObjectVariables?

4ian commented on newIDE/app/src/SceneEditor/index.js:
```diff
@@ -1352,37 +1372,6 @@ export default class SceneEditor extends React.Component<Props, State> {
             }}
           />
         )}
-        {!!this.state.variablesEditedObject && (
```
You also need to clean the unused `variablesEditedObject` in the state, the type and the now useless `editObjectVariables`. 
Otherwise, this will clutter the file and create fake complexity for the future developer working of the file.

blurymind commented:
This still keeps the instance variables available in the left side properties panel right? My only concern is in not adding a requirement for an extra click to get to it

4ian commented:
Yes, this just relocates the object variables from their own window into a tab in the object editor.
Instance variables are untouched :)

Midhil457 commented:
> This seems wrong, you check the object of the edited instance, but we're in the object editor, so we're editing an object.

instead, just pass this.state.editedObjectWithContext.object to the last argument of findAllObjectVariables?

It was `this.state.editedObjectWithContext.object` at first, but the test gave me an error at that line, after some testing, that was what solved it, I didn't see that it was calling instance variables

> You also need to clean the unused variablesEditedObject in the state, the type and the now useless editObjectVariables.
Otherwise, this will clutter the file and create fake complexity for the future developer working of the file.

I tried removing as many references that I thought were useless, so I might have missed some even this time


- Changed empty message to EmptyPlaceholderMessage
- Added props for setting Help Page Path
- The instance variable doc doesn't exist. I will create one in the next few weeks, but still have added the link as help path


Changes have been applied to:

- [x] Scene Variable Editor
- [x] Global Variable Editor
- [x] Object Variable Editor
- [x] Instance Variable Editor
- [x] Scene Variable Editor in Scene Variable Field
- [x] Object Variable Editor in Object Variable Field
- [x] Global Variable Editor in Global Variable Field

Bouh commented on newIDE/app/src/ObjectEditor/ObjectEditorDialog.js:
```diff
@@ -188,6 +196,36 @@ const InnerDialog = (props: InnerDialogProps) => {
         />
       )}
+      {currentTab === 'variables' && (
+        <Column expand noMargin>
```
`useFullHeight` is missing here?

Midhil457 replied:
It doesn't seem to help with the positioning issue
Done it like: `<Column expand noMargin useFullHeight>`

4ian commented on newIDE/app/src/EventsSheet/ParameterFields/GlobalVariableField.js:
```diff
@@ -64,10 +64,17 @@ export default class GlobalVariableField extends React.Component<
             }
+            emptyExplanationSecondMessage={
+              <Trans>
+                For example, you can have a variable called UnlockedLevelsCount
```
Thanks for making this consistent with the project manager 👍

4ian commented:
I believe I fixed the issues of centering, and also rework the VariablesList component and the variables editor dialog so that it's also centered here and the toolbar stay visible at all time :) 
Also a bunch of Flow typing fixes.

Should be good to go!

Object variables:
![image](https://user-images.githubusercontent.com/1280130/143143255-4ba22ab4-5111-4577-959e-3dd33b21c975.png)

Object variables (empty):
![image](https://user-images.githubusercontent.com/1280130/143143327-fcf57269-5184-45ae-9081-94ff37590a12.png)

Instance variables in the panel:
![image](https://user-images.githubusercontent.com/1280130/143143359-bff5ba6c-9a64-4def-9c15-a27091c04cb0.png)

Instance variables in their editor:
![image](https://user-images.githubusercontent.com/1280130/143143392-b7cec42c-165d-49b2-bd65-f4cd319a9e8d.png)

Instance variables in their editor (empty):
![image](https://user-images.githubusercontent.com/1280130/143143425-39e773bc-0e13-45db-aacd-e19fe01c5ef9.png)
